pkg/reconciler/knativeeventing/common/replicasenvvarstransform.go

```diff
@@ -53,13 +62,31 @@ func ReplicasEnvVarsTransform(client unstructuredGetter) mf.Transformer {
 			// Keep the existing number of replicas in the cluster for the deployment
 			apply.Spec.Replicas = current.Spec.Replicas
 
+			// Preserve the env vars in the existing cluster
 			for index := range current.Spec.Template.Spec.Containers {
 				currentContainer := current.Spec.Template.Spec.Containers[index]
-				for j := range apply.Spec.Template.Spec.Containers {
-					applyContainer := &apply.Spec.Template.Spec.Containers[j]
-					if currentContainer.Name == applyContainer.Name {
-						applyContainer.Env = currentContainer.Env
+				found, containerIndex := nameExistsContainers(currentContainer.Name, apply.Spec.Template.Spec.Containers)
+				if found {
+					applyContainer := &apply.Spec.Template.Spec.Containers[containerIndex]
+					var mergedEnv []corev1.EnvVar
+					for _, env := range currentContainer.Env {
+						if envVarNames.Has(env.Name) {
+							// Keep all the env vars in the preserved list
+							mergedEnv = append(mergedEnv, env)
+						}
 					}
+					for _, env := range applyContainer.Env {
+						if !envVarNames.Has(env.Name) {
+							// If the env var in the applied container is not in the preserved list,
+							// we should apply it.
+							mergedEnv = append(mergedEnv, env)
+						} else if !nameExistsEnvVars(env.Name, currentContainer.Env) {
```

Comment:
Why is this path necessary? We start fresh and have a clear preserve list, which we exclude already. We should append all other env vars here I believe.

Reply:
if the env belongs to the preserved list, the current does not have this env, but the apply container has it, we SHOULD apply this env. No?

Reply:
GREAT point! I guess the more performant way of doing that is to build up a new `sets.String` in the first loop, adding every variable that was **actually** there, and then check against that new list in the second loop. Spares you the quad complexity. Somthing like

```golang
actualKeys := sets.NewString()
for _, env := range currentContainer.Env {
    if envVarNames.Has(env.Name) {
        // Keep all the env vars in the preserved list
        mergedEnv = append(mergedEnv, env)
        actualKeys.Insert(env.Name)
    }
}

for _, env := range applyContainer.Env {
    if !actualKeys.Has(env.Name) {
        // Apply all keys that are neither preserved, nor in the actual container.
        mergedEnv = append(mergedEnv, env)
    }
}
```

I might be halluzinating though, since I'm super hungry :joy:. Feel free to ignore this comment too :), the performance implications aren't going to matter anyway.

Reply:
Feel free to call it a day. Thx for the comments.